astacus/node/cassandra.py

```diff
@@ -0,0 +1,156 @@
+                ipc.CassandraSubOp.remove_snapshot: self.remove_snapshot,
+                ipc.CassandraSubOp.restore_snapshot: self.restore_snapshot,
+                ipc.CassandraSubOp.start_cassandra: self.start_cassandra,
+                ipc.CassandraSubOp.stop_cassandra: self.stop_cassandra,
+                ipc.CassandraSubOp.take_snapshot: self.take_snapshot,
+            }[subop],
+        )
+
+    def _get_schema_hash(self):
+        with CassandraClient(self.config.cassandra.client).connect() as cas:
+            schema = CassandraSchema.from_cassandra_session(cas)
+            return schema.calculate_hash()
+
+    def get_schema_hash(self):
+        """This is used to get hash of the schema as seen by this node."""
+        with contextlib.suppress(TransientException):
+            self.result.schema_hash = self._get_schema_hash()
+        self.result.progress.done()
+
+    def remove_snapshot(self):
+        """This is used to remove the current snapshot (if any).
+
+        It is used as prelude for actual Astacus snapshot of the files
+        and after the backup has completed.
+
+        Note that Cassandra does not do any internal bookkeeping of
+        the snapshots so the rmtrees are enough.
+        """
+        progress = self.result.progress
+        progress.add_total(1)
+        todo = list(self.config.root.glob(SNAPSHOT_GLOB))
+        progress.add_success()
+        progress.add_total(len(todo))
+        for snapshotpath in todo:
+            shutil.rmtree(snapshotpath)
+            progress.add_success()
+        progress.done()
+
+    def restore_snapshot(self):
+        """This is used to restore the snapshot files into place, with Cassandra offline."""
+        # TBD: Delete extra data (current cashew doesn't do it, but we could)
+
+        # Move files from Astacus snapshot directories to the actual data directories
+        progress = self.result.progress
+        table_snapshots = list(self.config.root.glob(SNAPSHOT_GLOB))
+        progress.add_total(len(table_snapshots))
+
+        for table_snapshot in table_snapshots:
+            parts = table_snapshot.parts
+            # -2 = snapshots, -1 = name of the snapshots
+            table_name_and_id = parts[-3]
+            keyspace_name = parts[-4]
+            if is_system_keyspace(keyspace_name):
+                progress.add_success()
+                continue
+
+            table_name, _ = table_name_and_id.rsplit("-", 1)
+
+            # This could be more efficient too; oh well.
+            keyspace_path = table_snapshot.parents[2]
+            table_paths = list(keyspace_path.glob(f"{table_name}-*"))
+            assert len(table_paths) >= 1, f"NO tables with prefix {table_name}- found in {keyspace_path}!"
+            if len(table_paths) > 1:
+                # Prefer the one that isn't table_name_and_id
+                table_paths = [p for p in table_paths if p.name != table_name_and_id]
+            assert len(table_paths) == 1
```

Comment:
maybe also assert that there is no data in the directory we're about to move stuff? This could lead to _interesting_ consequences if someone managed to create 2 versions of the same table (during a network split, for example) and we ended up with 2 `tableX-uuid` paths in the snapshot. On restore, first we move stuff from `tableX-1` in the snapshot to `tableX-...` in the data dirs, then try to move stuff from `tableX-2`... I'd prefer we'd just fail the restoration in that case, seems very unlikely to happen in production. _Probably_ Cassandra includes only the active version of the table in the snapshot, but I'm not sure.

Reply:
See above - I think it may include two in some cases.

Reply:
I suppose that's fair - there seems to be backups/ subfolder by default but otherwise it should be empty when we start restoring it.